### ISSUE TITLE
Route goes cross-country as straight line for unreachable destination

### ISSUE THREAD

**Issue by osmonoid** [osmandapp/OsmAnd#16181]
### Routing engine
- [x] OsmAnd's in-app offline routing

### Routing options

profile: car
avoid: none
allow private access: off

### Start and end points

<table>
	<tr>
		<td></td>
		<td>route A</td>
		<td>route B</td>
		<td>route C</td>
	</tr>
	<tr>
		<td></td>
		<td>obvious; simplified example</td>
		<td>obvious; still quite simple; actual destination</td>
		<td>less obvious without zooming; problematic/realistic case</td>
	</tr>
	<tr>
		<td>departure</td>
		<td><a href="https://www.openstreetmap.org/?mlat=46.6822&mlon=12.1476#map=15/46.6822/12.1476">geo:46.6822,12.1476</a></td>
		<td><a href="https://www.openstreetmap.org/?mlat=46.6822&mlon=12.1476#map=15/46.6822/12.1476">geo:46.6822,12.1476</a></td>
		<td><a href="https://www.openstreetmap.org/?mlat=46.7164&mlon=11.6580#map=15/46.7164/11.6580">geo:46.7164,11.6580</a></td>
	</tr>
	<tr>
		<td>destination</td>
		<td><a href="https://www.openstreetmap.org/?mlat=46.6723&mlon=12.1520#map=15/46.6723/12.1520">geo:46.6723,12.1520</a></td>
		<td><a href="https://www.openstreetmap.org/?mlat=46.6567&mlon=12.1751#map=15/46.6567/12.1751">geo:46.6567,12.1751</a></td>
		<td><a href="https://www.openstreetmap.org/?mlat=46.6567&mlon=12.1751#map=15/46.6567/12.1751">geo:46.6567,12.1751</a></td>
	</tr>
</table>

### Actual routes
Route partially goes cross-country as a straight line not following any road/track (depending on the starting point: not meters, but kilometers long).

<table>
	<tr>
		<td>route A</td>
		<td>route B</td>
		<td>route C</td>
	</tr>
	<tr>
		<td><img src="https://user-images.githubusercontent.com/122121671/211150770-5eed7c86-000f-436c-8fd7-ec384154a564.jpg" width=108px height=240px></td>
		<td><img src="https://user-images.githubusercontent.com/122121671/211150628-12114fd7-266f-4b3e-84ad-8805dab7af84.jpg" width=125px height=240px></td>
		<td><img src="https://user-images.githubusercontent.com/122121671/211151202-93ef667e-4cca-45de-9c94-fb3edac408d4.jpg" width=501px height=240px></td>
	</tr> 
</table>

### Expected routes
Route (blue line) ends at point closest to the destination reachable with the selected means of transport (in this case car) since no valid route to the actual destination exists.
With the 🏁 flag still being at the unreachable destination and the route (blue line) prematurely ending before that, the user has at least a chance of noticing that something is wrong before starting the trip, reaching the problematic point on the way where the route first leads away from the correct highway, onto a (dirt) track, finally ending in the middle of nowhere (as it is the case for the problematic example route C).
Expected route C can be obtained e.g. by [OSRM FOSSGIS](https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=46.7164%2C11.6581%3B46.6567%2C12.1751#map=11/46.7474/11.9003):
![OSRM FOSSGIS](https://user-images.githubusercontent.com/122121671/211151969-f0a5c9d0-ad2a-4213-a673-f51d90ad5dd7.jpg)

### Expected routes "with cherry on top"
Same as above plus additionally a warning for the user that the selected destination is unreachable (at least with the selected means of transport) and that the computed route ends prematurely.

### Is this a regression?

don't know
first observed in version 4.2.7

## Environment

OsmAnd Version: 4.3.5
Device: Google Pixel 4a
Android version: 12
Maps used: offline - Italy_trentino-alto-adige_europe_2 2022-10-01

**Anything else relevant?**
The stumbling block for the bug to occur in this case is most likely [the barrier](https://www.openstreetmap.org/node/6569761700) which is not mapped correctly (conditional motor vehicle access is actually possible).
I deliberately did not fix the node attributes in the map data itelf yet, since it would prevent the reproduction of this issue.
Yes, I know, if the user selects a destination in the forest, far from any road, the result is similar: a straight line through the vegetation starting at the closest road. But in such a case the error is more obvious as the user has at least the chance to see the mistake (no road nearby) when selecting the destination from the map.
In contrast, if you pick "reasonable" start and end points and are routed through the wilderness, it is not as obvious, especially when the point of departure is further away: the straight line segment is not easily visible at high (above the ground) zoom levels.

**Comment by vshcherb** [comment]
There is no route through this barrier, so it can't go through - https://www.openstreetmap.org/node/6569761700 and it's considered to have no route.
https://www.openstreetmap.org/node/1362002825

**Comment by osmonoid** [comment]
Since you prematurely closed this issue I am not sure if I got the main point / reason for this issue across.

Yes, the destination is in fact (based on incorrect underlying OSM data) unreachable for cars.
I explicitly stated that this is clear in the title and the description.
As I also wrote that I deliberately did not change this mapping error in the data yet in order to get my main message through.

The point is, that OsmAnd routing pretends / acts as if it was reachable (please see the screenshots).

In contrast, the [OSRM FOSSGIS](https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=46.7164%2C11.6581%3B46.6567%2C12.1751#map=11/46.7474/11.9003) routing makes it clear at a single glance, that there is a problem with the selected destination.

Does this explanation make the actual problem more clear and make this issue eligible as an actual bug?

**Comment by vshcherb** [comment]
It's not a problem itself cause we have other requests to draw a straight line for unreachable destination and include it in total distance. Indeed I don't see a big issue whether it's indicated as straight line or as a stop before, so anyway the better UI is needed so we indicate "End point is not reachable"

**Comment by osmonoid** [comment]
> [...] we have other requests to draw a straight line for unreachable destination and include it in total distance [...]

I did not know that - do you have a link to the corresponding issue?

I started searching for it and by accident I came across another issue where (not the creator, but a different user) actually described a somewhat [very similar case](https://github.com/osmandapp/OsmAnd/issues/1633#issuecomment-1369915117) part of issue #1633.

<img src="https://user-images.githubusercontent.com/8058880/210390361-0c4daada-8282-4b39-b1a8-69e1094de2e0.png" width="180px" height="320px"></img>

What do you think about @sonora's [suggestion](https://github.com/osmandapp/OsmAnd/issues/1633#issuecomment-1369948964 ):

> [...] Maybe we should make that last segment more visually distinguishable to facilitate this. [...]

So instead of completely leaving out the problematic straight line part of the route (as I initially suggested not being aware of this other contradicting "straight line request" you mentioned) I propose a concrete alternative suggestion:

Make the straight line part stick out more clearly visible from the rest of the "correct" route by coloring it differently.
E.g. if the route is in blue, paint the straight line in red - see the following image edit of the problematic route (note the small red part at the end):

<img src="https://user-images.githubusercontent.com/122121671/212369073-d0fed051-f73f-4ab4-99dd-b2d0b5cebc48.jpg" width="320px" height="153px"></img>

Then you have at least the chance to see something is wrong when looking at the route on a a higher zoom level.

What do you think?

**Comment by goldfndr** [comment]
Alternative suggestion: another 🏁 flag (or other flag/icon) at the last reachable node.

**Comment by vshcherb** [comment]
Yes we have other similar and more critical issues like the navigation across the river, still to think to find out, ... etc. So I would consider this closed as duplciate
